pre_commit/languages/download.py

```diff
@@ -0,0 +1,258 @@
+
+    @property
+    def checksum(self) -> str:
+        return self._checksum
+
+    @property
+    def digest(self) -> str:
+        return b64encode(self._hasher.digest()).decode('utf-8')
+
+    def update(self, data: bytes) -> None:
+        return self._hasher.update(data)
+
+    def check_content(self) -> None:
+        if self.digest != self.checksum:
+            raise ContentCorruptionError(f"""SRI checker content corruption,
+the digest is {self.algorithm}-{self.digest}""")
+
+    def __str__(self) -> str:
+        return self.value
+
+
+@dataclass(frozen=True)
+class URI:
+    value: str
+
+    def __post_init__(self) -> None:
+        url = urlparse(self.value)
+        if not all([url.scheme, url.netloc]):
+            raise ValueError(f'Invalid URI: {self.value}')
+
+    def __str__(self) -> str:
+        return self.value
+
+
+@dataclass(frozen=True)
+class Metadata:
+    value: str
+
+    @property
+    def parts(self) -> tuple[str, str, str, str]:
+        first, second, third, fourth = self.value.splitlines()
+        return (first, second, third, fourth)
+
+    @property
+    def platform(self) -> Platform:
+        platform, _, _, _ = self.parts
+        return Platform(platform)
+
+    @property
+    def sri(self) -> SRI:
+        _, sri, _, _ = self.parts
+        return SRI(sri)
+
+    @property
+    def uri(self) -> URI:
+        _, _, uri, _ = self.parts
+        return URI(uri)
+
+    @property
+    def filename(self) -> PurePath:
+        _, _, _, path = self.parts
+        return PurePath(path)
+
+
+def io_buffer(io: IO[AnyStr], size: int = 4096) -> Iterator[AnyStr]:
+    """Read chunks from a stream.
+    :param io: the stream to read chunks from
+    :param size: the size of chunks to read
+    :returns: iterator of the chunks
+    """
+    buffer = io.read(size)
+    while buffer:
+        yield buffer
+        buffer = io.read(size)
+
+
+def download(uri: URI, sri: SRI, filename: Path) -> None:
+    filename.parent.mkdir(parents=True, exist_ok=True)
+    with urlopen(str(uri)) as ws:
+        with filename.open('wb') as fp:
+            for buf in io_buffer(ws):
+                fp.write(buf)
+                sri.update(buf)
+            fp.flush()
+    sri.check_content()
+    chmod(filename, stat.S_IRUSR | stat.S_IXUSR)
+
+
+def install_environment(
+        prefix: Prefix,
+        version: str,
+        additional_dependencies: Sequence[str],
+) -> None:
+    host = Platform.host()
+    for dep in additional_dependencies:
+        m = Metadata(dep)
+        if host == m.platform:
+            envdir = Path(
+                lang_base.environment_dir(
+                    prefix, ENVIRONMENT_DIR,
+                    version,
+                ),
+            )
+            filename = envdir / m.filename
+            download(m.uri, m.sri, filename)
+            srisum = envdir / 'health.srisum'
+            with srisum.open('w', encoding='utf8') as stream:
+                stream.write(f'{filename} {m.sri}\n')
```

Comment:
Sum files have the hash first, two spaces and then the filename:

```suggestion
                stream.write(f'{m.sri}  {filename}\n')
```

Reply:
Done